dissect/target/plugins/apps/edr/velociraptor.py

```diff
@@ -0,0 +1,106 @@
+import json
+import re
+import urllib.parse
+from collections.abc import Iterator
+from functools import lru_cache
+
+from flow.record import Record
+
+from dissect.target.exceptions import UnsupportedPluginError
+from dissect.target.helpers.record import DynamicDescriptor, TargetRecordDescriptor
+from dissect.target.plugin import Plugin, export
+from dissect.target.target import Target
+
+VELOCIRAPTOR_RESULTS = "/$velociraptor_results$"
+ISO_8601_PATTERN = r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?(Z|[+-]\d{2}:\d{2})?"
+
+
+class VelociraptorRecordBuilder:
+    def __init__(self, artifact_name: str):
+        self._create_event_descriptor = lru_cache(4096)(self._create_event_descriptor)
+        self.record_name = f"velociraptor/{artifact_name}"
+
+    def build(self, object: dict, target: Target) -> TargetRecordDescriptor:
+        """Builds a Velociraptor record."""
+        record_values = {}
+        record_fields = []
+
+        record_values["_target"] = target
+
+        for key, value in object.items():
+            # Reserved by flow.record
+            if key.startswith("_"):
+                continue
+
+            key = key.lower()
+
+            if re.match(ISO_8601_PATTERN, str(value)):
+                record_type = "datetime"
+            elif isinstance(value, list):
+                record_type = "string[]"
+            elif isinstance(value, int):
+                record_type = "varint"
+            elif key == "hash":
+                record_type = "digest"
+                value = (value.get("MD5"), value.get("SHA1"), value.get("SHA256"))
+            elif isinstance(value, str):
+                record_type = "string"
+            elif isinstance(value, dict):
+                record_type = "record"
+                # Lower key names in nested records
+                value = {k.lower(): v for k, v in value.items()}
```

Comment:
```suggestion
                value = self.build(value, target)
```
This is how a nested record should be serialized, but I'm not sure if that'll still work for your purpose?

Reply:
Hm not sure, perhaps we can argue for a `dict` type in `flow.record`. @yunzheng is currently on vacation though so that'll have to wait.

Another option is to flatten the dict to keys like `authenticode_programname`, `authenticode_issuername` etc.

Reply:
Okay I am fine with leaving the code as is with the nested records. I have created an issue (https://github.com/fox-it/dissect.target/issues/1128) for the support of the field type `dict`.